### PR TITLE
ocamlPackages.nbd: init at 1.20.2

### PR DESCRIPTION
**Pull request by lukts30**
Enable compilation of ocaml bindings in the main libnbd derivation, and expose them in ocamlPackages.
Needed for virt-v2v.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/top-level/ocaml-packages.nix [hunk 1]
```diff
@@ -976,6 +976,10 @@ let
 
     libc = callPackage ../development/ocaml-modules/libc { };
 
+    libnbd = pkgs.libnbd.override {
+      buildOcamlBindings = true;
```
Comment (vbgl) [inline pkgs/top-level/ocaml-packages.nix:980]:
```suggestion
    nbd = pkgs.libnbd.override {
      ocamlPackages = self;
      buildOcamlBindings = true;
```

**Comment by vbgl** [comment]
The OCaml bindings require a “recent” version of the compiler. Wrong configurations can be detected at evaluation time, e.g. by adding the following assertion:

~~~
--- a/pkgs/by-name/li/libnbd/package.nix
+++ b/pkgs/by-name/li/libnbd/package.nix
@@ -16,6 +16,9 @@
   autoreconfHook,
 }:

+lib.throwIf (buildOcamlBindings && !lib.versionAtLeast ocamlPackages.ocaml.version "4.05")
+  "OCaml binding are not available for OCaml < 4.05"
+
 stdenv.mkDerivation rec {
   pname = "libnbd";
   version = "1.20.2";
~~~

**Review by vbgl — APPROVED** [review]

File: pkgs/by-name/li/libnbd/package.nix [hunk 2]
```diff
@@ -1,16 +1,24 @@
...
+  libxml2,
+  fuse,
+  fuse3,
+  gnutls,
+  autoreconfHook,
 }:
 
+lib.throwIf (buildOcamlBindings && !lib.versionAtLeast ocamlPackages.ocaml.version "4.05")
+  "OCaml binding are not available for OCaml < 4.05"
+
```
Comment (SigmaSquadron) [inline pkgs/by-name/li/libnbd/package.nix:21]:
```suggestion
assert lib.assertMsg (buildOcamlBindings && !lib.versionAtLeast ocamlPackages.ocaml.version "4.05") "OCaml binding are not available for OCaml < 4.05";
  
```

Comment (vbgl) [reply pkgs/by-name/li/libnbd/package.nix:21]: Why? Don’t failing asserts break evaluation?

**Comment by SigmaSquadron** [comment]
My apologies, I assumed you wanted to fail evaluation here. Just a non-failing trace seems odd, and would just lead to a build failure if it triggered.

**Comment by vbgl** [comment]
The documentation for [`throw`](https://nix.dev/manual/nix/2.18/language/builtins.html#builtins-throw) highlights the difference with `assert`.

However, in practice, I cannot spot any run-time difference between before and after the change you propose. So my question stay: why do you suggest this change?

**Comment by SigmaSquadron** [comment]
Consistency with other packages, mostly. I haven't seen a library function just called before the actual derivation function like that; usually it's an assert that precedes derivation calls for handling... assertions.
Also, wouldn't the && operator fail if we aren't building OCaml? The logical implication seems to make more sense here IMO.

**Comment by vbgl** [comment]
There are a few instances of this pattern in the repo, e.g.: https://github.com/NixOS/nixpkgs/blob/5190ddc27dc6a083df998f50abb38557932d80b0/pkgs/games/dwarf-fortress/wrapper/default.nix#L144-L148

The condition you wrote is exactly the negation of the other one.

**Pull request merged**